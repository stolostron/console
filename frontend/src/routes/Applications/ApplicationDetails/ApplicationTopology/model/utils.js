/* Copyright Contributors to the Open Cluster Management project */
// Copyright (c) 2021 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project
/* eslint no-param-reassign: "error" */
import _ from 'lodash'
import { nodeMustHavePods } from '../helpers/diagram-helpers-utils'

const localClusterName = 'local-cluster'

export const getClusterName = (nodeId) => {
  if (nodeId === undefined) {
    return ''
  }
  const clusterIndex = nodeId.indexOf('--clusters--')
  if (clusterIndex !== -1) {
    const startPos = nodeId.indexOf('--clusters--') + 12
    const endPos = nodeId.indexOf('--', startPos)
    return nodeId.slice(startPos, endPos > 0 ? endPos : nodeId.length)
  }
  return localClusterName
}

export const createChildNode = (parentObject, clustersNames, type, links, nodes, replicaCount = 1) => {
  const parentType = _.get(parentObject, 'type', '')
  const { name, namespace, id, specs = {} } = parentObject
  const parentId = id
  const memberId = `${parentId}--${type}--${name}`
  let resources
  if (specs.resources) {
    resources = specs.resources.map((res) => {
      return { ...res, kind: type }
    })
  }
  const resourceCount = specs.resourceCount === 0 ? replicaCount : specs.resourceCount * replicaCount
  const node = {
    name,
    namespace,
    type,
    id: memberId,
    uid: memberId,
    specs: {
      isDesign: false,
      resourceCount,
      resources,
      clustersNames,
      replicaCount,
      parent: {
        parentId,
        parentName: name,
        parentType,
        resources: specs.resources,
        parentSpecs: _.get(specs, 'parent.parentSpecs'),
      },
    },
  }
  nodes.push(node)
  links.push({
    from: { uid: parentId },
    to: { uid: memberId },
    type: '',
  })
  return node
}

// add cluster node to RHCAM application
export const addClusters = (parentId, subscription, source, clusterNames, managedClusters, links, nodes, topology) => {
  // create element if not already created
  const sortedClusterNames = _.sortBy(clusterNames)
  let clusterId = 'member--clusters'
  // do not use this for the id for argo app, we only know about one app here
  if (subscription) {
    const cns = sortedClusterNames.join('--')
    const sub = _.get(subscription, 'metadata.name')
    clusterId = `member--clusters--${cns}--${sub}`
  } else {
    clusterId = 'member--clusters--'
  }
  const topoClusterNode = topology
    ? _.find(topology.nodes, {
        id: 'member--clusters',
      })
    : undefined
  //    const filteredClusters = managedClusterNames.filter((cluster) => {
  //        const cname = _.get(cluster, metadataName)
  //        return cname && clusterNames.includes(cname)
  //    })
  nodes.push({
    name: clusterNames.length === 1 ? clusterNames[0] : '',
    namespace: '',
    type: 'cluster',
    id: clusterId,
    uid: clusterId,
    specs: {
      title: source,
      subscription,
      resourceCount: clusterNames.length,
      //            cluster: subscription && filteredClusters.length === 1 ? filteredClusters[0] : undefined,
      //            clusters: filteredClusters,
      clustersNames: clusterNames,
      clusters: _.cloneDeep(managedClusters),
      sortedClusterNames,
      appClusters: topoClusterNode ? topoClusterNode.specs.appClusters : undefined,
      targetNamespaces: topoClusterNode ? topoClusterNode.specs.targetNamespaces : undefined,
    },
  })
  links.push({
    from: { uid: parentId },
    to: { uid: clusterId },
    type: '',
    specs: { isDesign: true },
  })
  return clusterId
}

export const getApplicationData = (nodes) => {
  let subscriptionName = ''
  let nbOfSubscriptions = 0
  let resourceMustHavePods = false
  const nodeTypes = []
  const result = {}
  let isArgoApp = false
  const appNode = nodes?.find((r) => r.type === 'application')
  if (appNode) {
    isArgoApp = _.get(appNode, ['specs', 'raw', 'apiVersion'], '').indexOf('argo') !== -1
    result.isArgoApp = isArgoApp
    //get argo app destination namespaces 'show_search':
    if (isArgoApp) {
      const applicationSetRef = _.get(appNode, ['specs', 'raw', 'metadata', 'ownerReferences'], []).find(
        (owner) => owner.apiVersion.startsWith('argoproj.io/') && owner.kind === 'ApplicationSet'
      )
      if (applicationSetRef) {
        result.applicationSet = applicationSetRef.name
      }
      let cluster = 'local-cluster'
      const clusterNames = _.get(appNode, ['specs', 'cluster-names'], [])
      if (clusterNames.length > 0) {
        cluster = clusterNames[0]
      }
      result.cluster = cluster
      result.source = _.get(appNode, ['specs', 'raw', 'spec', 'source'], {})
    }
  }
  nodes?.forEach((node) => {
    const type = _.get(node, 'type', '')
    const nodeType = type === 'project' ? 'namespace' : type
    if (!(isArgoApp && _.includes(['application', 'cluster'], nodeType))) {
      nodeTypes.push(nodeType) //ask for this related object type
    }
    if (nodeMustHavePods(node)) {
      //request pods when asking for related resources, this resource can have pods
      resourceMustHavePods = true
    }
    if (nodeType === 'subscription') {
      subscriptionName = _.get(node, 'name', '')
      nbOfSubscriptions = nbOfSubscriptions + 1
    }
  })

  if (resourceMustHavePods) {
    nodeTypes.push('pod')
  }

  //if only one subscription, ask for resources only related to that subscription
  result.subscription = nbOfSubscriptions === 1 ? subscriptionName : null
  //ask only for these type of resources since only those are displayed
  result.relatedKinds = _.uniq(nodeTypes)

  return result
}

export const getAppSetArgoCluster = (search, clusters) => {
  return clusters.find((cluster) => cluster.name === search || cluster.url === search)
}

export const processMultiples = (resources) => {
  if (resources.length > 5) {
    const groupByKind = _.groupBy(resources, 'kind')
    return Object.entries(groupByKind).map(([kind, _resources]) => {
      if (_resources.length === 1) {
        return _resources[0]
      } else {
        return {
          kind,
          name: '',
          namespace: '',
          resources: _resources,
          resourceCount: _resources.length,
        }
      }
    })
  }
  return resources
}
