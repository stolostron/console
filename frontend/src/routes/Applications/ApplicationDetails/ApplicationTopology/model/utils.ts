/* Copyright Contributors to the Open Cluster Management project */
/* eslint no-param-reassign: "error" */

import _ from 'lodash'
import { nodeMustHavePods } from '../helpers/diagram-helpers-utils'
import type { ApplicationData, ManagedCluster, Topology, TopologyLink, TopologyNode } from './types'

// Extract cluster name encoded in a node id; fall back to hub cluster if not encoded
export const getClusterName = (nodeId: string | undefined, hubClusterName?: string): string => {
  if (nodeId === undefined) {
    return ''
  }
  const clusterIndex = nodeId.indexOf('--clusters--')
  if (clusterIndex !== -1) {
    const startPos = nodeId.indexOf('--clusters--') + 12
    const endPos = nodeId.indexOf('--', startPos)
    return nodeId.slice(startPos, endPos > 0 ? endPos : nodeId.length)
  }
  return hubClusterName ?? ''
}

// Create a typed child node under a parent topology node and link them
export const createChildNode = (
  parentObject: TopologyNode,
  clustersNames: string[],
  type: string,
  links: TopologyLink[],
  nodes: TopologyNode[],
  replicaCount: number = 1
): TopologyNode => {
  const parentType = _.get(parentObject, 'type', '') as string
  const { name, namespace, id, specs = {} } = parentObject
  const parentId = id
  const memberId = `${parentId}--${type}--${name}`
  let resources: unknown
  if ((specs as Record<string, unknown>).resources) {
    resources = (specs as { resources: Array<Record<string, unknown>> }).resources.map((res) => {
      return { ...res, kind: type }
    })
  }
  const parentResourceCount = (specs as Record<string, unknown>).resourceCount as number | undefined
  const resourceCount = parentResourceCount === 0 ? replicaCount : (parentResourceCount ?? 1) * replicaCount
  const node: TopologyNode = {
    name,
    namespace,
    type,
    id: memberId,
    uid: memberId,
    specs: {
      isDesign: false,
      resourceCount,
      resources,
      clustersNames,
      replicaCount,
      parent: {
        parentId,
        parentName: name,
        parentType,
        resources: (specs as Record<string, unknown>).resources,
        parentSpecs: _.get(specs, 'parent.parentSpecs'),
      },
    },
  }
  nodes.push(node)
  links.push({
    from: { uid: parentId },
    to: { uid: memberId },
    type: '',
  })
  return node
}

// Add a synthetic "clusters" node under a parent and link them
export const addClusters = (
  parentId: string,
  subscription: Record<string, unknown> | undefined,
  source: string,
  clusterNames: string[],
  managedClusters: ManagedCluster[],
  links: TopologyLink[],
  nodes: TopologyNode[],
  topology?: Topology
): string => {
  // create element if not already created
  const sortedClusterNames = _.sortBy(clusterNames)
  let clusterId = 'member--clusters'
  // do not use this for the id for argo app, we only know about one app here
  if (subscription) {
    const cns = sortedClusterNames.join('--')
    const sub = _.get(subscription, 'metadata.name') as string | undefined
    clusterId = `member--clusters--${cns}--${sub}`
  } else {
    clusterId = 'member--clusters--'
  }
  const topoClusterNode = topology
    ? _.find(topology.nodes, {
        id: 'member--clusters',
      })
    : undefined
  nodes.push({
    name: clusterNames.length === 1 ? clusterNames[0] : '',
    namespace: '',
    type: 'cluster',
    id: clusterId,
    uid: clusterId,
    specs: {
      title: source,
      subscription,
      resourceCount: clusterNames.length,
      clustersNames: clusterNames,
      clusters: _.cloneDeep(managedClusters),
      sortedClusterNames,
      appClusters: topoClusterNode ? (topoClusterNode as TopologyNode).specs.appClusters : undefined,
      targetNamespaces: topoClusterNode ? (topoClusterNode as TopologyNode).specs.targetNamespaces : undefined,
    },
  })
  links.push({
    from: { uid: parentId },
    to: { uid: clusterId },
    type: '',
    specs: { isDesign: true },
  })
  return clusterId
}

// Collects info needed to query Search for related kinds from the visible nodes
export const getApplicationData = (nodes: TopologyNode[] | undefined, hubClusterName: string): ApplicationData => {
  let subscriptionName = ''
  let nbOfSubscriptions = 0
  let resourceMustHavePods = false
  const nodeTypes: string[] = []
  const result: ApplicationData = {
    subscription: null,
    relatedKinds: [],
  }
  let isArgoApp = false
  const appNode = nodes?.find((r) => r.type === 'application')
  if (appNode) {
    isArgoApp = _.get(appNode, ['specs', 'raw', 'apiVersion'], '').indexOf('argo') !== -1
    result.isArgoApp = isArgoApp
    // get argo app destination namespaces 'show_search'
    if (isArgoApp) {
      const applicationSetRef = _.get(appNode, ['specs', 'raw', 'metadata', 'ownerReferences'], []).find(
        (owner: Record<string, string>) =>
          owner.apiVersion.startsWith('argoproj.io/') && owner.kind === 'ApplicationSet'
      )
      if (applicationSetRef) {
        result.applicationSet = applicationSetRef.name
      }
      let cluster = hubClusterName
      const clusterNames = _.get(appNode, ['specs', 'clusterNames'], []) as string[]
      if (clusterNames.length > 0) {
        cluster = clusterNames[0]
      }
      result.cluster = cluster
      result.source = _.get(appNode, ['specs', 'raw', 'spec', 'source'], {}) as Record<string, unknown>
    }
  }
  nodes?.forEach((node) => {
    const type = _.get(node, 'type', '') as string
    const nodeType = type === 'project' ? 'namespace' : type
    if (!(isArgoApp && _.includes(['cluster'], nodeType))) {
      nodeTypes.push(nodeType) // ask for this related object type
    }
    if (nodeMustHavePods(node as unknown as Record<string, unknown>)) {
      // request pods when asking for related resources, this resource can have pods
      resourceMustHavePods = true
    }
    if (nodeType === 'subscription') {
      subscriptionName = _.get(node, 'name', '') as string
      nbOfSubscriptions = nbOfSubscriptions + 1
    }
  })

  if (resourceMustHavePods) {
    nodeTypes.push('pod')
  }

  // if only one subscription, ask for resources only related to that subscription
  result.subscription = nbOfSubscriptions === 1 ? subscriptionName : null
  // ask only for these type of resources since only those are displayed
  result.relatedKinds = _.uniq(nodeTypes)

  return result
}

// Find an ApplicationSet cluster by name or url
export const getAppSetArgoCluster = (search: string, clusters: ManagedCluster[]): ManagedCluster | undefined => {
  return clusters.find((cluster) => cluster.name === search || (cluster as unknown as { url?: string }).url === search)
}

// Group large lists of resources by kind for presentation; multiply counts by deployed clusters
export const processMultiples = (
  resources: Array<Record<string, unknown>>,
  numOfDeployedClusters: number = 0
): Array<Record<string, unknown>> => {
  // Need to multiply the number of success clusters
  const rCount = numOfDeployedClusters > 0 ? resources.length * numOfDeployedClusters : resources.length
  if (rCount > 5) {
    const groupByKind = _.groupBy(resources, 'kind')
    return Object.entries(groupByKind).map(([kind, _resources]) => {
      if (_resources.length === 1) {
        return _resources[0]
      } else {
        return {
          kind,
          name: '',
          namespace: '',
          resources: _resources,
          resourceCount: numOfDeployedClusters > 0 ? _resources.length * numOfDeployedClusters : _resources.length,
        }
      }
    })
  }
  return resources
}
