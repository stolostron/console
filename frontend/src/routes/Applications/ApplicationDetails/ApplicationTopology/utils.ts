/* Copyright Contributors to the Open Cluster Management project */

// Utility function to safely get nested properties (replaces lodash get)
export function safeGet<T = any>(obj: any, path: string | string[], defaultValue?: T): T {
  if (!obj || typeof obj !== 'object') return defaultValue as T

  const keys = Array.isArray(path) ? path : path.split(/[.[\]]+/).filter((key) => key !== '')
  let result = obj

  for (const key of keys) {
    if (result == null || typeof result !== 'object') {
      return defaultValue as T
    }
    result = result[key]
  }

  return result === undefined ? (defaultValue as T) : result
}

/**
 * Helper function to safely get nested properties from an object
 */
export const getNestedProperty = (obj: any, path: string | number | (string | number)[], defaultValue?: any): any => {
  if (obj == null) return defaultValue

  const pathArray = Array.isArray(path) ? path : String(path).split('.')
  let result = obj

  for (const key of pathArray) {
    if (result == null) return defaultValue
    result = result[key]
  }

  return result !== undefined ? result : defaultValue
}

// Utility function to safely set nested properties (replaces lodash set)
export function safeSet(obj: any, path: string | string[], value: any): void {
  if (!obj || typeof obj !== 'object') return

  const keys = Array.isArray(path) ? path : path.split('.')
  let current = obj

  for (let i = 0; i < keys.length - 1; i++) {
    const key = keys[i]
    if (!(key in current) || typeof current[key] !== 'object' || current[key] === null) {
      current[key] = {}
    }
    current = current[key]
  }

  current[keys[keys.length - 1]] = value
}

// Deep clone utility (replaces lodash cloneDeep)
export function deepClone<T>(obj: T): T {
  if (obj === null || typeof obj !== 'object') return obj
  if (obj instanceof Date) return new Date(obj.getTime()) as unknown as T
  if (obj instanceof Array) return obj.map((item) => deepClone(item)) as unknown as T
  if (typeof obj === 'object') {
    const cloned = {} as T
    for (const key in obj) {
      if (Object.prototype.hasOwnProperty.call(obj, key)) {
        cloned[key] = deepClone(obj[key])
      }
    }
    return cloned
  }
  return obj
}

/**
 * Deep equality comparison for objects
 * @param obj1 - First object to compare
 * @param obj2 - Second object to compare
 * @returns true if objects are deeply equal, false otherwise
 */
export function deepEqual(obj1: any, obj2: any): boolean {
  if (obj1 === obj2) return true
  if (obj1 == null || obj2 == null) return false
  if (typeof obj1 !== 'object' || typeof obj2 !== 'object') return false

  const keys1 = Object.keys(obj1)
  const keys2 = Object.keys(obj2)

  if (keys1.length !== keys2.length) return false

  for (const key of keys1) {
    if (!keys2.includes(key)) return false
    if (!deepEqual(obj1[key], obj2[key])) return false
  }

  return true
}
