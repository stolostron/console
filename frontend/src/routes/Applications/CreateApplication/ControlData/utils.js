export const getUniqueChannelName = (channelPath, groupControlData) => {
    //create a unique name for a new channel, based on path and type
    if (!channelPath || !groupControlData) {
      return ''
    }
  
    //get the channel type and append to url to make sure different type of channels are unique, yet using the same url
    const channelTypeSection = groupControlData.find(
      ({ id }) => id === 'channelType'
    )
  
    let channelTypeStr
    let channelType
    if (channelTypeSection) {
      channelTypeStr = _.get(channelTypeSection, 'active', [''])[0]
    }
  
    switch (channelTypeStr) {
    case 'github':
      channelType = 'g'
      break
    case 'helmrepo':
      channelType = 'h'
      break
    case 'objectstore':
      channelType = 'o'
      break
    default:
      channelType = 'ns'
    }
  
    let channelName = _.trim(channelPath)
    if (_.startsWith(channelName, 'https://')) {
      channelName = _.trimStart(channelName, 'https://')
    }
    if (_.startsWith(channelName, 'http://')) {
      channelName = _.trimStart(channelName, 'http://')
    }
    if (_.endsWith(channelName, '.git')) {
      channelName = _.trimEnd(channelName, '.git')
    }
  
    channelName = _.replace(channelName, /\./g, '')
    channelName = _.replace(channelName, /:/g, '')
    channelName = _.replace(channelName, /\//g, '-')
  
    channelName = _.trimEnd(channelName, '-')
    channelName = channelName.toLowerCase()
  
    //max name for ns or resources is 63 chars
    // trim channel name to max 58 char to allow a max of 63 char length
    //for the channel authentication (which is channelName-auth) object and channel ns (channelName-ns)
    if (channelName.length > 58) {
      channelName = channelName.substring(channelName.length - 56)
    }
    channelName = `${channelType}${channelName}`
  
    return channelName
  }