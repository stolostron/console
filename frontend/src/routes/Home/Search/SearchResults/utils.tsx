/* Copyright Contributors to the Open Cluster Management project */
import queryString from 'query-string'
import { TFunction } from 'react-i18next'
import { generatePath, useHistory } from 'react-router-dom'
import { NavigationPath } from '../../../../NavigationPath'
import { ClosedDeleteModalProps, IDeleteModalProps } from '../components/Modals/DeleteResourceModal'
import { GetUrlSearchParam } from '../searchDefinitions'

export interface ISearchResult {
  kind: string
  apiversion: string
  apigroup?: string
  __type: string
}

export function GetRowActions(
  kind: string,
  currentQuery: string,
  relatedResource: boolean,
  setDeleteResource: React.Dispatch<React.SetStateAction<IDeleteModalProps>>,
  t: TFunction
) {
  const history = useHistory()

  const viewApplication = {
    id: 'view-application',
    title: t('View application'),
    click: (item: any) => {
      const { apigroup, applicationSet, cluster, name, namespace, kind } = item
      if (apigroup === 'app.k8s.io' || apigroup === 'argoproj.io') {
        const params = queryString.stringify({
          apiVersion: `${kind}.${apigroup}`.toLowerCase(),
          cluster: cluster === 'local-cluster' ? undefined : cluster,
          applicationset: applicationSet ?? undefined,
        })
        return history.push({
          pathname: generatePath(NavigationPath.applicationOverview, {
            namespace,
            name,
          }),
          search: `?${params}`,
          state: {
            from: NavigationPath.search,
            fromSearch: window.location.search,
          },
        })
      }
      const searchParams = GetUrlSearchParam(item)
      return history.push({
        pathname: NavigationPath.resourceRelated,
        search: searchParams,
        state: {
          from: NavigationPath.search,
          fromSearch: window.location.search,
        },
      })
    },
  }
  const viewAppTopology = {
    id: 'view-application-topology',
    title: t('View application topology'),
    click: (item: any) => {
      const apiversion = encodeURIComponent(`${item?.kind}.${item?.apigroup}`.toLowerCase())
      return history.push({
        pathname: generatePath(NavigationPath.applicationTopology, { name: item?.name, namespace: item?.namespace }),
        search: `?apiVersion=${apiversion}`,
        state: {
          from: NavigationPath.search,
          fromSearch: window.location.search,
        },
      })
    },
  }
  const editButton = {
    id: 'edit',
    title: t('Edit {{resourceKind}}', { resourceKind: kind }),
    click: (item: any) => {
      const searchParams = GetUrlSearchParam(item)
      return history.push({
        pathname: NavigationPath.resourceYAML,
        search: searchParams,
        state: {
          from: NavigationPath.search,
          fromSearch: window.location.search,
        },
      })
    },
  }
  const viewRelatedButton = {
    id: 'view-related',
    title: t('View related resources'),
    click: (item: any) => {
      const searchParams = GetUrlSearchParam(item)
      return history.push({
        pathname: NavigationPath.resourceRelated,
        search: searchParams,
        state: {
          from: NavigationPath.search,
          fromSearch: window.location.search,
        },
      })
    },
  }
  const deleteButton = {
    id: 'delete',
    title: t('Delete {{resourceKind}}', { resourceKind: kind }),
    click: (item: any) => {
      setDeleteResource({
        open: true,
        close: () => setDeleteResource(ClosedDeleteModalProps),
        resource: item,
        currentQuery,
        relatedResource,
      })
    },
  }

  if (kind === 'cluster' || kind === 'release' || kind === 'policyreport') {
    return []
  } else if (kind === 'application') {
    return [viewApplication, viewAppTopology, editButton, viewRelatedButton, deleteButton]
  }
  return [editButton, viewRelatedButton, deleteButton]
}
