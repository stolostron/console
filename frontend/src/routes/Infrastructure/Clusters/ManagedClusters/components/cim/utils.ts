/* Copyright Contributors to the Open Cluster Management project */
import isEqual from 'lodash/isEqual'
import { CIM } from 'openshift-assisted-ui-lib'

const appendPatch = (patches: any, path: string, newVal: object | string, existingVal?: object | string) => {
    if (!isEqual(newVal, existingVal)) {
        patches.push({
            op: existingVal ? 'replace' : 'add',
            path,
            value: newVal,
        })
    }
}

export const getNetworkingPatches = (agentClusterInstall: CIM.AgentClusterInstallK8sResource, values: any) => {
    const agentClusterInstallPatches: any = []

    appendPatch(
        agentClusterInstallPatches,
        '/spec/sshPublicKey',
        values.sshPublicKey,
        agentClusterInstall.spec.sshPublicKey
    )

    appendPatch(
        agentClusterInstallPatches,
        '/spec/networking/clusterNetwork',
        [
            {
                cidr: values.clusterNetworkCidr,
                hostPrefix: values.clusterNetworkHostPrefix,
            },
        ],
        agentClusterInstall.spec?.networking?.clusterNetwork
    )

    appendPatch(
        agentClusterInstallPatches,
        '/spec/networking/serviceNetwork',
        [values.serviceNetworkCidr],
        agentClusterInstall.spec?.networking?.serviceNetwork
    )

    // Setting Machine network CIDR is forbidden when cluster is not in vip-dhcp-allocation mode (which is not soppurted ATM anyway)
    if (values.vipDhcpAllocation) {
        const hostSubnet = values.hostSubnet?.split(' ')?.[0]
        const machineNetworkValue = hostSubnet ? [{ cidr: hostSubnet }] : []
        appendPatch(
            agentClusterInstallPatches,
            '/spec/networking/machineNetwork',
            machineNetworkValue,
            agentClusterInstall.spec?.networking?.machineNetwork
        )
    }

    appendPatch(agentClusterInstallPatches, '/spec/apiVIP', values.apiVip, agentClusterInstall.spec?.apiVIP)

    appendPatch(agentClusterInstallPatches, '/spec/ingressVIP', values.ingressVip, agentClusterInstall.spec?.ingressVIP)

    return agentClusterInstallPatches
}
