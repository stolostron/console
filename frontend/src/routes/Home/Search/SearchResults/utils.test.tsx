// Copyright (c) 2023 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project
import i18next from 'i18next'
import { getSearchDefinitions } from '../searchDefinitions'
import { generateSearchResultExport, GetRowActions } from './utils'

const mockHistoryPush = jest.fn()
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}))

const t = i18next.t.bind(i18next)

test('Correctly return row Actions', () => {
  const res = GetRowActions('Pod', 'kind:Pod', false, () => {}, t)
  res[0].click({ kind: 'Pod' }) // edit resource
  res[1].click({ kind: 'Pod' }) // view related resources
  res[2].click({ kind: 'Pod' }) // delete resource
  expect(res).toMatchSnapshot()
})

test('Correctly return empty row Actions for restricted resource', () => {
  const res = GetRowActions('Cluster', 'kind:Cluster', false, () => {}, t)
  expect(res).toMatchSnapshot()
})

test('Correctly return empty row Actions for Application', () => {
  const res = GetRowActions('Application', 'kind:Application', false, () => {}, t)
  res[0].click({
    apigroup: 'app.k8s.io',
    kind: 'Application',
    cluster: 'local-cluster',
    name: 'testApp',
    namespace: 'testAppNs',
  }) // edit app
  res[1].click({
    apigroup: 'app.k8s.io',
    kind: 'Application',
    cluster: 'local-cluster',
    name: 'testApp',
    namespace: 'testAppNs',
  }) // view app topology
  expect(res).toMatchSnapshot()
})

test('generateSearchResultExport - Correctly generates and triggers csv download for single resource kind', () => {
  const toastContextMock: any = {
    addAlert: jest.fn(),
  }
  const t = (key: string) => key
  window.URL.createObjectURL = jest.fn()

  const searchResultDataMock = {
    searchResult: [
      {
        items: [
          {
            apigroup: 'operators.coreos.com',
            apiversion: 'v1alpha1',
            channel: 'release-2.11',
            cluster: 'local-cluster',
            created: '2024-04-15T14:20:26Z',
            installplan: 'advanced-cluster-management.v2.11.0',
            kind: 'Subscription',
            kind_plural: 'subscriptions',
            label: 'operators.coreos.com/advanced-cluster-management.open-cluster-management=',
            name: 'acm-operator-subscription',
            namespace: 'open-cluster-management',
            package: 'advanced-cluster-management',
            phase: 'AtLatestKnown',
            source: 'acm-custom-registry',
          },
        ],
        __typename: 'SearchResult',
      },
    ],
  }

  const searchDefinitions = getSearchDefinitions((key) => key)
  generateSearchResultExport(searchResultDataMock, searchDefinitions, toastContextMock, t)

  expect(toastContextMock.addAlert).toHaveBeenCalledWith({
    title: 'Generating data. Download may take a moment to start.',
    type: 'info',
    autoClose: true,
  })
  expect(window.URL.createObjectURL).toHaveBeenCalledTimes(1)
  expect(toastContextMock.addAlert).toHaveBeenCalledWith({
    title: 'Export successful',
    type: 'success',
    autoClose: true,
  })
})

test('generateSearchResultExport - Correctly generates and triggers csv download for multi resource kind', () => {
  const toastContextMock: any = {
    addAlert: jest.fn(),
  }
  const t = (key: string) => key
  window.URL.createObjectURL = jest.fn()

  const searchResultDataMock = {
    searchResult: [
      {
        items: [
          {
            apigroup: 'operators.coreos.com',
            apiversion: 'v1alpha1',
            channel: 'release-2.11',
            cluster: 'local-cluster',
            created: '2024-04-15T14:20:26Z',
            installplan: 'advanced-cluster-management.v2.11.0',
            kind: 'Subscription',
            kind_plural: 'subscriptions',
            label: 'operators.coreos.com/advanced-cluster-management.open-cluster-management=',
            name: 'acm-operator-subscription',
            namespace: 'open-cluster-management',
            package: 'advanced-cluster-management',
            phase: 'AtLatestKnown',
            source: 'acm-custom-registry',
          },
          {
            _hubClusterResource: 'true',
            _ownerUID: 'local-cluster/1234-abcd',
            _uid: 'local-cluster/1234-abcd',
            apiversion: 'v1',
            cluster: 'local-cluster',
            container: 'search-api',
            created: '2024-04-15T14:23:59Z',
            hostIP: '10.0.3.162',
            image: 'quay.io/image',
            kind: 'Pod',
            kind_plural: 'pods',
            label: 'app=search; component=search-v2-operator; name=search-api; pod-template-hash=69775dc595',
            name: 'search-api-69775dc595-lhrk9',
            namespace: 'open-cluster-management',
            podIP: '10.130.0.114',
            restarts: '0',
            startedAt: '2024-04-15T14:23:59Z',
            status: 'Running',
          },
          {
            _hubClusterResource: 'true',
            _ownerUID: 'local-cluster/1234-abcd',
            _uid: 'local-cluster/1234-abcd',
            apiversion: 'v1',
            cluster: 'local-cluster',
            container: 'search-collector',
            created: '2024-04-15T14:23:59Z',
            hostIP: '10.0.68.86',
            image: 'quay.io/image',
            kind: 'Pod',
            kind_plural: 'pods',
            name: 'search-collector-764d748c4f-pbgrb',
            namespace: 'open-cluster-management',
            podIP: '10.129.0.117',
            restarts: '0',
            startedAt: '2024-04-15T14:23:59Z',
            status: 'Running',
          },
        ],
        __typename: 'SearchResult',
      },
    ],
  }

  const searchDefinitions = getSearchDefinitions((key) => key)
  generateSearchResultExport(searchResultDataMock, searchDefinitions, toastContextMock, t)

  expect(toastContextMock.addAlert).toHaveBeenCalledWith({
    title: 'Generating data. Download may take a moment to start.',
    type: 'info',
    autoClose: true,
  })
  expect(window.URL.createObjectURL).toHaveBeenCalledTimes(1)
  expect(toastContextMock.addAlert).toHaveBeenCalledWith({
    title: 'Export successful',
    type: 'success',
    autoClose: true,
  })
})
