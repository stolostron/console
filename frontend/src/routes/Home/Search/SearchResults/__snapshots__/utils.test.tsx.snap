// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correctly return empty row Actions for Application 1`] = `
Array [
  Object {
    "click": [Function],
    "id": "view-application",
    "title": "View application",
  },
  Object {
    "click": [Function],
    "id": "view-application-topology",
    "title": "View application topology",
  },
  Object {
    "click": [Function],
    "id": "edit",
    "title": "Edit application",
  },
  Object {
    "click": [Function],
    "id": "view-related",
    "title": "View related resources",
  },
  Object {
    "click": [Function],
    "id": "delete",
    "title": "Delete application",
  },
]
`;

exports[`Correctly return empty row Actions for restricted resource 1`] = `Array []`;

exports[`Correctly return row Actions 1`] = `
Array [
  Object {
    "click": [Function],
    "id": "edit",
    "title": "Edit pod",
  },
  Object {
    "click": [Function],
    "id": "view-related",
    "title": "View related resources",
  },
  Object {
    "click": [Function],
    "id": "delete",
    "title": "Delete pod",
  },
]
`;
